esp-hal/src/lib.rs

```diff
@@ -644,3 +635,18 @@ macro_rules! before_snippet {
         ))
     };
 }
+
+use crate::{
+    clock::{ClockControl, Clocks, CpuClock},
+    peripherals::Peripherals,
+};
+
+/// Initialize the system.
+///
+/// This function sets up the CPU clock and returns the peripherals and clocks.
+pub fn init(clock_config: CpuClock) -> (Peripherals, Clocks<'static>) {
+    let peripherals = Peripherals::take();
+    let clocks = ClockControl::new(clock_config).freeze();
```

Comment:
If we plan to never support dynamic clocks, which IMO could be a good idea, we could perhaps remove `Clocks` from the API completely, and keep it as an implementation detail within the HAL.

Reply:
Definitely, and definitely as part of this PR. I'd love to not introduce another 1000 lines of change at once.

Reply:
I suppose one thing to consider, is that ancillary crates often need to know the configured clocks, esp-wifi, esp-hal-smartled, and the new esp-hal-buzzer. So we still need some way of obtaining this information. I guess this could be as simple as `esp_hal::clocks::Clocks::get()` or something like that.

Reply:
I'd say keep the `Clocks` struct at least for the sake of simplifying error handling. i.e. What happens if you call `esp_hal::clocks::Clocks::get()` before the clocks are initialized. The `Clocks` struct statically proves that they've been initalised.

Reply:
You'll not be able to work with peripherals without calling the init function, which sets up the clocks. The only exception right now is `Peripherals::take` which I plan on removing.

Otherwise, you'll need unsafe code to steal the peripherals and you're on your own at that point.